### ISSUE TITLE
Performance regression in 0.5.0

### ISSUE THREAD

**Issue by Byron** [jessegrosjean/jwalk#24]
Thanks for jwalk, it is literally the reason [dua](https://github.com/Byron/dua-cli/issues/41) is what it is.

However, there seems to be a performance regression in version 0.5.0. 
It can be reproduced as follows:

```sh
git clone https://github.com/Byron/dua-cli && \
cd dua-cli && \
git checkout 4990fa4202f2b687ee2476efe0a406fdfe23fd96 && \
cargo run --release -- i ~/
```

The CPU usage is indicative - it starts out using many cores, but quickly degrades to just a single core, as if work stealing doesn't work anymore or works differently. I saw this behaviour on MacOS using Rustc 1.42.

The latest master, which uses jwalk 0.4.0, does not have that issue.

The reason I would like to upgrade is the option of not following symlinks, and the more consistent API.

**Comment by jessegrosjean** [comment]
Bryon, and thanks for using jwalk... it's been fun trying to keep a public crate going... not something I've done much before. But I just merged that last (typo) pull request already! :) Hope to keep doing better.

I think this regression is due to API changes. Previously jwalk had a build in option to preload metadata when it was making parallel traversal. That's been removed because you can do the same thing with client state... but it's removal also means that now `dir_entry.metadata()` in `dua`'s code isn't cached... so the traversal is still being done as fast as before I think. But then any call to `dir_entry.metadata()` in `dua` code is hitting disk.

Please see this test case for how to load metadata into `client_state`:

https://github.com/jessegrosjean/jwalk/blob/master/benches/walk_benchmark.rs#L45

I recommend defining `client_state` type to hold just what you need for dua and then fill in that state from `process_read_dir` callback as in the benchmark.

And please let me know how it goes!

**Comment by Byron** [comment]
Thanks so much, @jessegrosjean, you were exactly right about the cause of the issue, and the example could essentially be copy-pasted to [upgrade my code](https://github.com/Byron/dua-cli/blob/b7ed2bbc957c416e8af08983bba46a4fe2a9553c/src/common.rs#L171) and do the job.

Dua was just released in v2.3.6, containing a slew of improvements and I think it's faster in interactive mode, too!
